### PR TITLE
Added PGV constraint for serverName.

### LINKED ISSUE #12709: [http_conn_manage]missing protobuf validation for config.serverName()
In the line
`headers.setReferenceServer(connection_manager_.config_.serverName());`
(the code is here: https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_impl.cc#L1824)

It adds the serverName of config to headers. However, header doesn't allow `\0\r\n` here:

```
static inline bool validHeaderString(absl::string_view s) {
  // If you modify this list of illegal embedded characters you will probably
  // want to change header_map_fuzz_impl_test at the same time.
  for (const char c : s) {
    switch (c) {
    case '\0':
      FALLTHRU;
    case '\r':
      FALLTHRU;
    case '\n':
      return false;
    default:
      continue;
    }
  }
  return true;
}
```

We may consider adding PGV constraint that the `serverName` field should not contain `\0\r\n`, or strip them when configuration.

### PR DESCRIPTION
**Pull request by mk46**
Signed-off-by: Manish Kumar <manish.kumar1@india.nec.com>

Commit Message: Added PGV constraint for serverName. It shouldn't contain \0\n\r. 
Additional Description: Added PGV constraint to not allow \0\n\r in serverName.Currently, header doesn't allow \0\r\n.
Risk Level:
Testing: config_test
Docs Changes:
Release Notes:
 Fixes #12709

### CODE REVIEW

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 1]
```diff
@@ -298,7 +298,7 @@ message HttpConnectionManager {
 
   // An optional override that the connection manager will write to the server
   // header in responses. If not set, the default is *envoy*.
-  string server_name = 10;
+  string server_name = 10 [(validate.rules).string = {pattern: "^[^\000\r\n]*$"}];
```
Comment (asraa) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:301]: nit: there is a pattern for this already `HTTP_HEADER_VALUE`

Comment (mk46) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:301]: I agree but the configuration should also have such constraints. Currently, server_name allow \0, \n, and \r which should not be there.

Comment (htuch) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:301]: I think @asraa is suggesting to use this pattern in the validator, see https://github.com/envoyproxy/envoy/blob/b6c8bb3a4ac6bcce221643a4924befd5eefd6815/api/envoy/config/core/v3/health_check.proto#L90 for an example.

Comment (mk46) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto]: @htuch @asraa Thanks for the review. PTAL.

**Review by asraa — APPROVED** [review]
Thanks for the test as well!

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**